Add GDALViewshedGenerate binding

pericles-tpt opened the pull request:
- implement `GDALViewshedGenerate` binding
- implement `Viewshed` go function (calls binding)
- add `TestViewshedSimpleHeight` to test the binding

pericles-tpt commented:
### Question: Invalid test results across GDAL versions

I found in my testing that different versions of GDAL pass different sections of the test:
1. 3.1.0 <    version < 3.4.2  - doesn't pass any tests
2. 3.4.2 <= version < 3.10.1 - passes all tests EXCEPT when output mode is DEM
3. 3.10.1+                             - passes all tests

My solution for this is to return an error for any case where tests produce an invalid result, the current behaviour for the above is:

1. Returns an error at the start of `Viewshed()`
2. Returns an error IF the output mode is DEM
3. Runs `Viewshed()` as normal

Do you prefer the approach I've implemented, which prevents the user from running `Viewshed()` when it could produce invalid data? Or, should I add a warning in the `Viewshed()` documentation comment and remove the errors for cases (1) and (2), this would allow the user to run `Viewshed()` for GDAL version 3.1.0+ at their own risk?

pericles-tpt commented:
@tbonfort I'm not sure if you're being notified about new PRs, but I'd really appreciate it if you could review this PR when you get the chance

tbonfort commented on godal.cpp:
```diff
@@ -1905,6 +1905,22 @@ GDALDatasetH godalDem(cctx *ctx, const char *pszDest, const char *pszProcessing,
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(3, 1, 0)
+	ret = GDALViewshedGenerate(bnd, pszDriverName, pszTargetRasterName, papszCreationOptions, dfObserverX, dfObserverY, dfObserverHeight, dfTargetHeight, dfVisibleVal, dfInvisibleVal, dfOutOfRangeVal, 
+	dfNoDataVal, dfCurvCoeff, GDALViewshedMode(eMode), dfMaxDistance, nullptr, nullptr, GDALViewshedOutputType(heightMode), nullptr);
+#endif
```
```suggestion
#else
	CPLError(CE_Failure, CPLE_AppDefined, "Viewshed not implemented in gdal < 3.1");
#endif
```

tbonfort commented on godal.go:
```diff
@@ -4060,6 +4060,88 @@ func (ds *Dataset) Dem(destPath, processingMode string, colorFilename string, sw
+	visibleVal float64, invisibleVal float64, outOfRangeVal float64, noDataVal float64, curveCoeff float64, mode ViewshedMode, maxDistance float64,
+	heightMode ViewshedOutputType, opts ...ViewshedOption) (*Dataset, error) {
+
+	// TODO: Should I put a 'warning' in the documentation for `Viewshed` instead of disallowing the last two configurations?
+	if !CheckMinVersion(3, 1, 0) {
+		return nil, errors.New("failed to run, 'viewshed' not supported on GDAL versions < 3.1.0")
+	} else if !CheckMinVersion(3, 4, 2) {
+		return nil, errors.New("cannot run 'viewshed' with GDAL version <= 3.4.1, as some tests produce invalid results under these conditions")
+	} else if !CheckMinVersion(3, 10, 0) && heightMode == MinTargetHeightFromDem {
+		return nil, errors.New("height mode CANNOT be `MinTargetHeightFromDem` when running a GDAL version < 3.10, as some tests produce invalid results under these conditions")
+	}
```
I would just leave this section out, it is not godal's role to circumvent fixed bugs in gdal. A comment with a link to the gdal PR/issue numbers is sufficient I believe

tbonfort commented on godal_test.go:
```diff
@@ -4440,6 +4440,179 @@ func TestGridInvalidSwitch(t *testing.T) {
 
+// Test Ported from: https://github.com/OSGeo/gdal/blob/6cdae8b8f7d09ecf67e24959e984d2e7bbe3ee62/autotest/cpp/test_viewshed.cpp#L98
+func TestViewshedSimpleHeight(t *testing.T) {
+	if !CheckMinVersion(3, 1, 0) {
+		_, err := Viewshed(Band{}, nil, "none", 1, 1, 0, 0, 255, 0, 0, -1, 0, MEdge, 0, Normal)
+		assert.EqualError(t, err, "failed to run, 'viewshed' not supported on GDAL versions < 3.1.0")
+		return
+	} else if !CheckMinVersion(3, 4, 2) {
+		_, err := Viewshed(Band{}, nil, "none", 1, 1, 0, 0, 255, 0, 0, -1, 0, MEdge, 0, Normal)
+		assert.EqualError(t, err, "cannot run 'viewshed' with GDAL version <= 3.4.1, as some tests produce invalid results under these conditions")
+		return
+	}
```
these versions of gdal are too old, no need to test for this.
```suggestion

```

tbonfort commented on godal_test.go:
```diff
@@ -4440,6 +4440,179 @@ func TestGridInvalidSwitch(t *testing.T) {
+}
+
+func TestViewshedCreationOptions(t *testing.T) {
+	if !CheckMinVersion(3, 1, 0) {
+		_, err := Viewshed(Band{}, nil, "none", 1, 1, 0, 0, 255, 0, 0, -1, 0, MEdge, 0, Normal)
+		assert.EqualError(t, err, "failed to run, 'viewshed' not supported on GDAL versions < 3.1.0")
+		return
+	} else if !CheckMinVersion(3, 4, 2) {
+		_, err := Viewshed(Band{}, nil, "none", 1, 1, 0, 0, 255, 0, 0, -1, 0, MEdge, 0, Normal)
+		assert.EqualError(t, err, "cannot run 'viewshed' with GDAL version <= 3.4.1, as some tests produce invalid results under these conditions")
+		return
+	}
```
these versions of gdal are too old, no need to test for this.
```suggestion
```

tbonfort reviewed (changes requested):
Sorry, I had missed the notifications. LGTM after these few nitpicks.

pericles-tpt commented:
@tbonfort thank you for reviewing the PR!

I’ve addressed most of your suggestions but I had to make some additional changes, they are:

1. I had to add back the “min version < 3.1.0” error check in my tests, as the ‘Go 1.23 + GDAL on ubuntu-20.04’ test seems to use a version of GDAL older than 3.1.0. I think you'd need to remove that test in order for me to remove the “min version < 3.1.0” check.

2. I also had to add a “min version >= 3.4.2” check to the 1st and 3rd scope in `TestViewshedSimpleHeight()` because I removed the version checks from `Viewshed()`

tbonfort commented on godal.go:
```diff
@@ -4060,6 +4060,88 @@ func (ds *Dataset) Dem(destPath, processingMode string, colorFilename string, sw
+//   - 3.1.0 <= version < 3.4.2: all tests fail
+//   - 3.4.2 <= version < 3.10.0: tests where heightMode == MinTargetHeightFromDem fail
+//   - version >= 3.10.0: all tests pass
+func Viewshed(targetBand Band, driverName *DriverName, targetRasterName string, observerX float64, observerY float64, observerHeight float64, targetHeight float64,
```
Sorry, I should have caught this in the first review:

I am not satisfied with the API you have come up with, as it directly maps the GDAL C API, and does not try to be "idiomatic go":

- `func (srcBand Band) Viewshed(...)`: should be a method on a Band receiver. rename targetBand to srcBand

- These are optional and/or have default values, so should also be passed in as a `ViewshedOption`:
  - driverName (reuse the exisiting one, same as for ErrHandler)
  - targetHeight
  - visibleVal (255) / invisibleVal (0) (could/should be passed in with a single ViewshedOption)
  - from what I understand outOfRangeVal and maxDistance can only go together, and thus should be passed in together as a single option
  - curveCoef
  - for noDataVal, mode and heightMode I don't have sufficient knowledge to determine if they should be considered as optional or explicitely required. if using an option, check if there is already an exisiting one for nodata

pericles-tpt replied:
@tbonfort I've pushed changes in https://github.com/airbusgeo/godal/pull/142/commits/46da29d404a613232778f034327bf58eec33bf7b to address the above, with some things to note:

-  I did not combine outOfRangeVal and maxDistance into a single option as I wasn't able to find a case where they could "only go together". I think it's valid to modify maxDistance on its own, as the default outOfRangeVal (i.e. 0) could still be valid if the other `Val` properties are non-zero values.
- From looking at the the `Options` provided to `GDALViewshedGenerate`, noDataVal, mode and heightMode all have default values so I've changed them to options as well
- I've also renamed the `mode` to `cellMode` as it's also referenced as that in GDAL and it more clearly communicates its function

tbonfort commented:
ubuntu 20.04 runner removed in #143

tbonfort commented on godal_test.go:
```diff
@@ -4440,6 +4440,159 @@ func TestGridInvalidSwitch(t *testing.T) {
+	assert.Error(t, err)
+
+	// Valid
+	_, err = vrtDs.Bands()[0].Viewshed("none", 2, 2, 0, CreationOption("TILED=YES", "BLOCKXSIZE=128", "BLOCKYSIZE=128"))
```
the returned dataset should be closed here to avoid a memory leak

tbonfort commented on godal.go:
```diff
@@ -4060,6 +4060,86 @@ func (ds *Dataset) Dem(destPath, processingMode string, colorFilename string, sw
+//
+// for more information see: https://gdal.org/en/stable/api/gdal_alg.html#_CPPv420GDALViewshedGenerate15GDALRasterBandHPKcPKc12CSLConstListddddddddd16GDALViewshedModed16GDALProgressFuncPv22GDALViewshedOutputType12CSLConstList
+//
+// Creations options can be set through options with:
+//
+//	Viewshed(bnd, "mem", "none", ..., CreationOption("TILED=YES","BLOCKXSIZE=256"))
+//
```
replace by list of available `ViewshedOption`s so they show up in IDE doc hints

pericles-tpt commented:
I've pushed 2 commits that address the "potential memory leaks" and "documentation comment missing ViewshedOptions" issues
